datafusion/core/src/execution/session_state.rs

```diff
@@ -976,6 +837,482 @@ impl SessionState {
+
+    /// Adds a new [`ConfigExtension`] to TableOptions
+    pub fn with_table_options_extension<T: ConfigExtension>(
+        mut self,
+        extension: T,
+    ) -> Self {
+        self.state.table_options.extensions.insert(extension);
+        self
+    }
+
+    /// override default execution props with `execution_props`
+    pub fn with_execution_props(mut self, execution_props: ExecutionProps) -> Self {
+        self.state.execution_props = execution_props;
+        self
+    }
+
+    /// override default table factories with `table_factories`
+    pub fn with_table_factories(
+        mut self,
+        table_factories: HashMap<String, Arc<dyn TableProviderFactory>>,
+    ) -> Self {
+        self.state.table_factories = table_factories;
+        self
+    }
+
+    /// override the runtime env with `runtime_env`
+    pub fn with_runtime_env(mut self, runtime_env: Arc<RuntimeEnv>) -> Self {
+        self.state.runtime_env = runtime_env;
+        self
+    }
+
+    /// Registers a [`FunctionFactory`] to handle `CREATE FUNCTION` statements
+    pub fn with_function_factory(
+        mut self,
+        function_factory: Option<Arc<dyn FunctionFactory>>,
+    ) -> Self {
+        self.state.function_factory = function_factory;
+        self
+    }
+
+    /// build a [`SessionState`] with the current configuration
+    pub fn build(mut self) -> SessionState {
+        if self.use_defaults {
+            if self.state.table_factories.is_empty() {
+                self.state.table_factories =
+                    SessionStateDefaults::default_table_factories();
+            }
+            if self.state.expr_planners.is_empty() {
+                self.state.expr_planners = SessionStateDefaults::default_expr_planners();
+            }
+
+            if self.state.config.create_default_catalog_and_schema() {
+                let default_catalog = SessionStateDefaults::default_catalog(
+                    &self.state.config,
+                    &self.state.table_factories,
+                    &self.state.runtime_env,
+                );
+
+                self.state.catalog_list.register_catalog(
+                    self.state.config.options().catalog.default_catalog.clone(),
+                    Arc::new(default_catalog),
+                );
+            }
+
+            if self.state.file_formats.is_empty() {
+                SessionStateDefaults::register_default_file_formats(&mut self.state);
+            }
+
+            if self.state.scalar_functions.is_empty() {
+                SessionStateDefaults::register_scalar_functions(&mut self.state);
+                SessionStateDefaults::register_array_functions(&mut self.state);
+            }
+
+            if self.state.aggregate_functions.is_empty() {
+                SessionStateDefaults::register_aggregate_functions(&mut self.state);
+            }
+        }
+
+        self.state.clone()
+    }
+}
+
+/// Defaults that are used as part of creating a SessionState such as table providers,
```

Comment:
This is great. Thank you

Eventually I would love to move this structure out of `session_state.rs` and into its own module / .rs file to make the distinction between SessionState and the default features clearer. However, we can do that as a follow on task too

Eventually I am thinking that we could pass `SessionStateDefaults` to the builder somehow (and thus finally break the dependency of SessionState on datafusion-core (but let's not try and do that in this PR)

Reply:
I didn't expect this PR to be approved as is, it was just to get feedback. I'm find with changing things to suit the future requirements.

Reply:
I thought it was good enough to merge and iterate on :) -- my thinking was that since we just released 40.0.0 we have a month to get it into shape before the next release. I am also happy to work on the initial PR too


Thank you